### PR TITLE
refactor: remove identifier.js ts-expect-error

### PR DESCRIPTION
**Pull request by SoonIter**
## Summary

https://github.com/web-infra-dev/rspack/issues/4640

### CODE REVIEW

**Comment by SoonIter** [comment]
😄 I just found that there will be so many vulnerabilities in webpack codes during migrating to ts

File: packages/rspack/src/util/identifier.js
Comment (chenjiahan) [inline packages/rspack/src/util/identifier.js:1]: Can we just turn `identifier.js` to `identifier.ts`?

File: packages/rspack/src/util/identifier.ts [hunk 1]
```diff
@@ -187,13 +193,13 @@ const makeCacheableWithContext = fn => {
...
 		 * @returns {string} the returned relative path
 		 */
-		const boundFn = (context, identifier) => {
-			let cachedResult;
-			// @ts-expect-error
-			let innerSubCache = innerCache.get(context);
+		const boundFn = (context: string, identifier: string): string => {
+			let cachedResult: string | undefined;
+			let innerSubCache: Map<string, string> | undefined =
+				innerCache && innerCache.get(context);
```
Comment (chenjiahan) [inline packages/rspack/src/util/identifier.ts:199]:
```suggestion
				innerCache?.get(context);
```

File: packages/rspack/src/util/identifier.ts [hunk 2]
```diff
@@ -187,13 +193,13 @@ const makeCacheableWithContext = fn => {
...
 			if (innerSubCache === undefined) {
-				// @ts-expect-error
-				innerCache.set(context, (innerSubCache = new Map()));
+				innerSubCache = new Map();
+				innerCache && innerCache.set(context, innerSubCache);
```
Comment (chenjiahan) [inline packages/rspack/src/util/identifier.ts:202]:
```suggestion
				innerCache?.set(context, innerSubCache);
```

File: packages/rspack/src/util/identifier.ts [hunk 3]
```diff
@@ -237,15 +245,13 @@ const makeCacheableWithContext = fn => {
 		 * @param {string} identifier identifier used to create relative path
 		 * @returns {string} the returned relative path
 		 */
-		const boundFn = identifier => {
-			// @ts-expect-error
-			const cachedResult = innerSubCache.get(identifier);
+		const boundFn = (identifier: string): string => {
+			const cachedResult = innerSubCache && innerSubCache.get(identifier);
```
Comment (chenjiahan) [inline packages/rspack/src/util/identifier.ts:249]:
```suggestion
			const cachedResult = innerSubCache?.get(identifier);
```

File: packages/rspack/src/util/identifier.ts [hunk 4]
```diff
@@ -237,15 +245,13 @@ const makeCacheableWithContext = fn => {
...
 			if (cachedResult !== undefined) {
 				return cachedResult;
 			} else {
 				const result = fn(context, identifier);
-				// @ts-expect-error
-				innerSubCache.set(identifier, result);
+				innerSubCache && innerSubCache.set(identifier, result);
```
Comment (chenjiahan) [inline packages/rspack/src/util/identifier.ts:254]:
```suggestion
				innerSubCache?.set(identifier, result);
```

**Review by chenjiahan — APPROVED** [review]

**Comment by chenjiahan** [comment]
👍🏻

**Pull request merged**